### PR TITLE
#2809: disable caching of NEW_CLASS

### PR DESCRIPTION
**Pull request by abslamp**
AnnotatedTypeFactory.fromExpressionTreeCache will return types with outdated type variables when visiting new operators. Disable caching of NEW_CLASS could be a solution for #2809.

### CODE REVIEW

File: framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 1]
```diff
@@ -1345,7 +1345,11 @@ private AnnotatedTypeMirror fromExpression(ExpressionTree tree) {
 
         AnnotatedTypeMirror result = TypeFromTree.fromExpression(this, tree);
 
-        if (shouldCache) {
+        if (shouldCache && tree.getKind() != Tree.Kind.NEW_CLASS) {
+            // Don't cache types of NEW operator
+            // Cached NEW_CLASS types may contain "Unknowned*" annotations
```
Comment (smillst) [inline framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: The cached types may contain incorrect annotations that were computed during dataflow.  The better solution is to only clear the cache if dataflow is running.

Comment (wmdietl) [reply framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: @smillst Thanks for the pointer to #602!
@abslamp Can you look into disabling caching during dataflow analysis and fixing #602?

Comment (abslamp) [reply framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: I will change the comments for now, but I want to look into the better solution.

**Comment by abslamp** [comment]
TODO: disable cache in dataflow

File: framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 2]
```diff
@@ -1345,7 +1345,11 @@ private AnnotatedTypeMirror fromExpression(ExpressionTree tree) {
...
+            // Such annotations are from previous dataflow analysis
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: @abslamp Can you update this PR and change this comment as suggested in https://github.com/typetools/checker-framework/pull/2643#discussion_r343972843

File: checker/tests/interning/Issue2809.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
+// test cases for #2809
+// https://github.com/typetools/checker-framework/issues/2809
+
+import org.checkerframework.checker.interning.qual.Interned;
+
+class Issue2809 {
```
Comment (wmdietl) [inline checker/tests/interning/Issue2809.java]: Could you add an expected failure test, to show that things that should go wrong still cause an error?

Comment (abslamp) [reply checker/tests/interning/Issue2809.java]: Do you mean a false negative one or just a normal expected failure?

**Pull request merged**